### ISSUE TITLE
ConnectionResetByPeer using MultiWorkerMirroredStrategy with Native Keras BERT Model

### ISSUE THREAD

**Issue by Eric-Le-Ge** [tensorflow/tensorflow#42498]
**Describe the current behavior**

I am trying to use the following script to run multi-worker distributed training on a Google Kubernetes Engine (GKE) cluster. The script completes and uploads the trained model when there is only one worker. However, it fails with a connection reset by peer error after some arbitrary number of steps when I use multiple nodes.

**Standalone code to reproduce the issue**

Training script: https://gist.github.com/Eric-Le-Ge/21fe4df4c8e361c60d20a683b08aa743
Original example: https://github.com/tensorflow/tfx/tree/master/tfx/examples/bert/cola

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Container-Optimized OS (cos) (default)
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.3.0
- Python version: 3.6.9
- Docker Image: tensorflow/tensorflow:latest-gpu
- GPU model and memory: NVIDIA K80

**Other info / logs** 
Pod execution logs:
https://gist.github.com/Eric-Le-Ge/693640219f54cb6350e2ea1cc4d97547

**Comment by nikitamaia** [comment]
Hi @Eric-Le-Ge, looking at the logs it's not clear to me that the training started. Where are you seeing that?

This seems like it will be a bit tricky to reproduce. For starters, are you seeing this happen every time you try to kick off multi worker training on GKE? If so, can you try out a simpler model to help determine if the problem is with TF or GKE?

Additionally, MultiWorkerMirroredStrategy can be run on a single machine without extra setup. Can you confirm that you're able to train successfully on each machine individually? If so, can you provide the logs for the successful training?

**Comment by Eric-Le-Ge** [comment]
Hi @nikitamaia , thank you for following up on this issue. I've tried training another model with the same setup and it ran successfully, so I think the issue is related to this specific model. In addition, I have successfully trained this model using a single node with the same script, and here are the logs from the worker:
https://gist.github.com/Eric-Le-Ge/48380f2c84afa635989dfc82d175d03a

**Comment by nikitamaia** [comment]
So to clarify, every time you try and run this model with two nodes you get this exact same error `Aborting RingReduce with Unavailable: Connection reset by peer`?
For the model that you were able to train successfully, was it also from tf hub?
I can try to repro this on my end. Can you provide complete instructions with data and how to run the script?

**Comment by nikitamaia** [comment]
Actually one more question -- for the single case it looks like the collective communication was set to RING. And it was AUTO when you trained with two nodes. This shouldn't make a difference, but just to sanity check can you also confirm that the two nodes fail with RING set as well.

**Comment by Eric-Le-Ge** [comment]
Thanks for following up! Yes, every time I run the script I get the exact same error, this also happens with 4 nodes as well. The other model is directly imported from `tf.keras.applications.MobileNet` and this model does not have the collective communication error. 
Also, when I got the original error the collective communication was set to AUTO. I tried setting it to RING and NCCL and both fail as well.

I think the error should be relatively easy to reproduce, but it requires setup for running TFX pipelines. The original example is here: https://github.com/tensorflow/tfx/tree/master/tfx/examples/bert/cola. The fastest way to reproduce this is to:
1. Run the pipeline locally and collect the data artifacts from the output of the Transform component. 
2. Pack the data artifacts with the training script into a container image.
3. Start a GKE cluster with 2 nodes, and set up TF Config for the nodes to communicate with each other.
4. Run the run_fn in both containers where fn_args contains the path to the transformed data.

Let me know if there is anything I can provide. Thank you!

**Comment by nikitamaia** [comment]
Do you face the same problem when running the original bert model with MWMS?

**Comment by Eric-Le-Ge** [comment]
Not sure I have not only tried running the BERT model provided in the example.

**Comment by nikitamaia** [comment]
To clarify, you are using the same `bert_cola_pipeline.py` and `bert_cola_pipeline_e2e_test.py` scripts from the `tfx/tfx/examples/bert/cola/` repo, and just swapping out the `bert_cola_utils.py` file?

**Comment by Eric-Le-Ge** [comment]
The original `bert_cola_pipeline.py` file runs on a single node so I only followed this pipeline up to the trainer part. By original COLA model do you mean the one provided in `bert_cola_utils.py`? If so the modified `bert_cola_utils.py` should be still running the same model.

**Comment by nikitamaia** [comment]
> I think the error should be relatively easy to reproduce, but it requires setup for running TFX pipelines. The original example is here: https://github.com/tensorflow/tfx/tree/master/tfx/examples/bert/cola. The fastest way to reproduce this is to:
> 
> 1. Run the pipeline locally and collect the data artifacts from the output of the Transform component.
> 2. Pack the data artifacts with the training script into a container image.
> 3. Start a GKE cluster with 2 nodes, and set up TF Config for the nodes to communicate with each other.
> 4. Run the run_fn in both containers where fn_args contains the path to the transformed data.
> 

I'm still a little confused by your above instructions to reproduce the issue. You linked to the original example but it seems that is not needed to reproduce, correct? All I need is [this bert_cola.utils.py](https://gist.github.com/Eric-Le-Ge/21fe4df4c8e361c60d20a683b08aa743) script from your repo?

**Comment by Eric-Le-Ge** [comment]
Yes, the original example is not needed to reproduce this issue, but the training steps requires transformed data and running the original example up to the Transform component would produce this data.

**Comment by ckkuang** [comment]
I suspect this is caused by instance preemption. MultiworkerMirroredStrategy requires all workers to be up to do gradient aggregation. Can you try using [BackupAndRestore](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/experimental/BackupAndRestore) callback, and wraps `model.fit` with a loop and try/catch [UnavailableError](https://www.tensorflow.org/api_docs/python/tf/errors/UnavailableError)?
